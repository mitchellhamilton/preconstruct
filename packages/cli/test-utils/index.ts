import path from "path";
import * as fs from "fs-extra";
import globby from "globby";
import fixturez from "fixturez";
import spawn from "spawndamnit";
// @ts-ignore
import initHasher from "xxhash-wasm";
// @ts-ignore
import profiler from "v8-profiler-next";

let f = fixturez(__dirname);

require("chalk").enabled = false;
console.error = jest.fn();
console.log = jest.fn();

export let logMock = {
  log: (console.log as any) as jest.MockInstance<void, any>,
  error: (console.error as any) as jest.MockInstance<void, any>
};

afterEach(() => {
  logMock.log.mockReset();
  logMock.error.mockReset();
});

import init from "../src/init";
import { confirms } from "../src/messages";

let mockedConfirms = confirms as jest.Mocked<typeof confirms>;

export async function initBasic(directory: string) {
  mockedConfirms.writeMainField.mockReturnValue(Promise.resolve(true));
  mockedConfirms.writeModuleField.mockReturnValue(Promise.resolve(true));

  await init(directory);
  mockedConfirms.writeMainField.mockReset();
  mockedConfirms.writeModuleField.mockReset();
}

export function profile(name: string) {
  profiler.startProfiling(name, true);

  return () => {
    let profile = profiler.stopProfiling();

    new Promise<void>(resolve =>
      profile
        .export()
        .pipe(fs.createWriteStream(name + ".cpuprofile"))
        .on("finish", function() {
          profile.delete();
          resolve();
        })
    );
  };
}

function getPkgPath(tmpPath: string) {
  return path.join(tmpPath, "package.json");
}

export async function getPkg(
  filepath: string
): Promise<{ [key: string]: any }> {
  return JSON.parse(
    await fs.readFile(path.join(filepath, "package.json"), "utf-8")
  );
}

export async function modifyPkg(
  tmpPath: string,
  cb: (pkgJson: { [key: string]: any }) => unknown
) {
  let json = await getPkg(tmpPath);
  await cb(json);

  let pkgPath = getPkgPath(tmpPath);
  await fs.writeFile(pkgPath, JSON.stringify(json, null, 2));
}

export let createPackageCheckTestCreator = (
  doResult: (tmpPath: string) => Promise<void>
) => {
  let createTestCreator = (
    testFn: (testName: string, func: () => Promise<void>) => void
  ) => async (
    testName: string,
    entrypoints: { [key: string]: any },
    cb: (
      doThing: () => Promise<{ [key: string]: any }>,
      stuff: { dir: string }
    ) => Promise<void>
  ) => {
    testFn(testName, async () => {
      let tmpPath = f.copy("template-simple-package");
      let things = Object.keys(entrypoints);
      await Promise.all(
        things.map(async entrypointPath => {
          let content = entrypoints[entrypointPath];
          let filepath = path.join(tmpPath, entrypointPath, "package.json");
          await fs.ensureFile(filepath);
          await fs.writeFile(filepath, JSON.stringify(content, null, 2));
        })
      );

      await cb(
        async () => {
          await doResult(tmpPath);

          let newThings: Record<string, any> = {};

          await Promise.all(
            things.map(async entrypointPath => {
              newThings[entrypointPath] = JSON.parse(
                await fs.readFile(
                  path.join(tmpPath, entrypointPath, "package.json"),
                  "utf8"
                )
              );
            })
          );
          return newThings;
        },
        { dir: tmpPath }
      );
    });
  };
  let testFn = createTestCreator(test);
  // @ts-ignore
  testFn.only = createTestCreator(test.only);
  // @ts-ignore
  testFn.skip = createTestCreator(test.skip);
  return testFn as typeof testFn & { only: typeof testFn; skip: typeof testFn };
};

export async function snapshotDistFiles(tmpPath: string) {
  let distPath = path.join(tmpPath, "dist");
  let distFiles;
  try {
    distFiles = await fs.readdir(distPath);
  } catch (err) {
    if (err.code === "ENOENT") {
      throw new Error(distPath + " does not exist");
    }
    throw err;
  }

  await Promise.all(
    distFiles.map(async x => {
      expect(
        await fs.readFile(path.join(distPath, x), "utf-8")
      ).toMatchSnapshot(x);
    })
  );
}

export let stripHashes = async (chunkName: string) => {
  let { h64 } = await initHasher();

  let transformer = (pathname: string, content: string) => {
    return pathname.replace(new RegExp(`${chunkName}-[^\\.]+`, "g"), () => {
      return `chunk-this-is-not-the-real-hash-${h64(content)}`;
    });
  };
  return {
    transformPath: transformer,
    transformContent: (content: string) => {
      return content.replace(new RegExp(`${chunkName}-[^\\.]+`, "g"), () => {
        return "chunk-some-hash";
      });
    }
  };
};

export async function snapshotDirectory(
  tmpPath: string,
  {
    files = "js",
    filterPath = x => true,
    transformPath = x => x,
    transformContent = x => x
  }: {
    files?: "all" | "js";
    filterPath?: (path: string) => boolean;
    transformPath?: (path: string, contents: string) => string;
    transformContent?: (content: string) => string;
  } = {}
) {
  let paths = await globby(
    [`**/${files === "js" ? "*.js" : "*"}`, "!node_modules/**", "!yarn.lock"],
    {
      cwd: tmpPath
    }
  );

  await Promise.all(
    paths.filter(fp => filterPath(fp)).map(async x => {
      let content = transformContent(
        await fs.readFile(path.join(tmpPath, x), "utf-8")
      );
      if (x.endsWith(".json") && !x.endsWith("tsconfig.json")) {
        content = JSON.parse(content);
      }
      expect(content).toMatchSnapshot(transformPath(x, content));
    })
  );
}

export async function install(tmpPath: string) {
  await spawn("yarn", ["install"], { cwd: tmpPath });
}
