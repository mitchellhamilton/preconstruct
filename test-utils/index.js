// @flow
import path from "path";
import * as fs from "fs-extra";
import globby from "globby";
import fixturez from "fixturez";

let f = fixturez(__dirname);

require("chalk").enabled = false;
// $FlowFixMe
console.error = jest.fn();
// $FlowFixMe
console.log = jest.fn();

export let logMock = {
  log: ((console.log: any): JestMockFn<any, void>),
  error: ((console.error: any): JestMockFn<any, void>)
};

afterEach(() => {
  logMock.log.mockReset();
  logMock.error.mockReset();
});

import init from "../src/init";
import { confirms } from "../src/messages";

export async function initBasic(directory: string) {
  confirms.writeMainField.mockReturnValue(true);
  confirms.writeModuleField.mockReturnValue(true);
  confirms.writeUmdBuilds.mockReturnValue(false);

  await init(directory);
  confirms.writeMainField.mockReset();
  confirms.writeModuleField.mockReset();
  confirms.writeUmdBuilds.mockReset();
}

function getPkgPath(tmpPath: string) {
  return path.join(tmpPath, "package.json");
}

export async function getPkg(filepath: string): Object {
  return JSON.parse(
    await fs.readFile(path.join(filepath, "package.json"), "utf-8")
  );
}

export async function modifyPkg(tmpPath: string, cb: Object => mixed) {
  let json = await getPkg(tmpPath);
  await cb(json);

  let pkgPath = getPkgPath(tmpPath);
  await fs.writeFile(pkgPath, JSON.stringify(json, null, 2));
}

export let createPackageCheckTestCreator = (
  doResult: string => Promise<void>
) => {
  let createTestCreator = testFn => async (
    testName: string,
    entrypoints: { [key: string]: Object },
    cb: (doThing: () => Promise<{ [key: string]: Object }>) => Promise<void>
  ) => {
    testFn(testName, async () => {
      let tmpPath = f.copy("template-simple-package");
      let things = Object.keys(entrypoints);
      await Promise.all(
        things.map(async entrypointPath => {
          let content = entrypoints[entrypointPath];
          let filepath = path.join(tmpPath, entrypointPath, "package.json");
          await fs.ensureFile(filepath);
          await fs.writeFile(filepath, JSON.stringify(content, null, 2));
        })
      );

      await cb(async () => {
        await doResult(tmpPath);

        let newThings = {};

        await Promise.all(
          things.map(async entrypointPath => {
            newThings[entrypointPath] = JSON.parse(
              await fs.readFile(
                path.join(tmpPath, entrypointPath, "package.json"),
                "utf8"
              )
            );
          })
        );
        return newThings;
      });
    });
  };
  let testFn = createTestCreator(test);
  testFn.only = createTestCreator(test.only);
  testFn.skip = createTestCreator(test.skip);
  return testFn;
};

export async function snapshotDistFiles(tmpPath: string) {
  let distPath = path.join(tmpPath, "dist");
  let distFiles;
  try {
    distFiles = await fs.readdir(distPath);
  } catch (err) {
    if (err.code === "ENOENT") {
      throw new Error(distPath + " does not exist");
    }
    throw err;
  }

  await Promise.all(
    distFiles.map(async x => {
      expect(
        await fs.readFile(path.join(distPath, x), "utf-8")
      ).toMatchSnapshot(x);
    })
  );
}

export async function snapshotDirectory(
  tmpPath: string,
  files: "all" | "js" = "js"
) {
  let paths = await globby([`**/${files === "js" ? "*.js" : "*"}`], {
    cwd: tmpPath
  });

  await Promise.all(
    paths.map(async x => {
      expect(await fs.readFile(path.join(tmpPath, x), "utf-8")).toMatchSnapshot(
        x
      );
    })
  );
}
