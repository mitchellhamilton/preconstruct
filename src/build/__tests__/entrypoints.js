// @flow
import build from "../";
import fixturez from "fixturez";
import { snapshotDistFiles, snapshotDirectory } from "../../../test-utils";

const f = fixturez(__dirname);

jest.mock("../../prompt");

test("source entrypoint option", async () => {
  let tmpPath = f.copy("source-entrypoint-option");

  await build(tmpPath);

  await snapshotDistFiles(tmpPath);
});

test("source entrypoint option flow", async () => {
  let tmpPath = f.copy("source-entrypoint-option-flow");

  await build(tmpPath);

  await snapshotDistFiles(tmpPath);
});

test("multiple entrypoints", async () => {
  let tmpPath = f.copy("multiple-entrypoints");

  await build(tmpPath);

  await snapshotDirectory(tmpPath);
});

test("two entrypoints, one module, one not", async () => {
  let tmpPath = f.copy("two-entrypoints-one-module-one-not");

  await build(tmpPath);

  await snapshotDirectory(tmpPath);
});

test("two entrypoints with a common dependency", async () => {
  let tmpPath = f.copy("common-dependency-two-entrypoints");

  await build(tmpPath);

  await snapshotDirectory(tmpPath);
});
