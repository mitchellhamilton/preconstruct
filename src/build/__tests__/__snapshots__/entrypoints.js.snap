// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multiple entrypoints: dist/multiple-entrypoints.cjs.dev.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

let sum = (a, b) => a + b;

exports.sum = sum;
"
`;

exports[`multiple entrypoints: dist/multiple-entrypoints.cjs.js 1`] = `
"'use strict';

if (process.env.NODE_ENV === \\"production\\") {
  module.exports = require(\\"./multiple-entrypoints.cjs.prod.js\\");
} else {
  module.exports = require(\\"./multiple-entrypoints.cjs.dev.js\\");
}
"
`;

exports[`multiple entrypoints: dist/multiple-entrypoints.cjs.prod.js 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: !0
});

let sum = (a, b) => a + b;

exports.sum = sum;
"
`;

exports[`multiple entrypoints: dist/multiple-entrypoints.esm.js 1`] = `
"let sum = (a, b) => a + b;

export { sum };
"
`;

exports[`multiple entrypoints: multiply/dist/multiple-entrypoints.cjs.dev.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

let multiply = (a, b) => a * b;

exports.multiply = multiply;
"
`;

exports[`multiple entrypoints: multiply/dist/multiple-entrypoints.cjs.js 1`] = `
"'use strict';

if (process.env.NODE_ENV === \\"production\\") {
  module.exports = require(\\"./multiple-entrypoints.cjs.prod.js\\");
} else {
  module.exports = require(\\"./multiple-entrypoints.cjs.dev.js\\");
}
"
`;

exports[`multiple entrypoints: multiply/dist/multiple-entrypoints.cjs.prod.js 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: !0
});

let multiply = (a, b) => a * b;

exports.multiply = multiply;
"
`;

exports[`multiple entrypoints: multiply/dist/multiple-entrypoints.esm.js 1`] = `
"let multiply = (a, b) => a * b;

export { multiply };
"
`;

exports[`multiple entrypoints: src/multiply.js 1`] = `"export let multiply = (a, b) => a * b;"`;

exports[`multiple entrypoints: src/sum.js 1`] = `"export let sum = (a, b) => a + b;"`;

exports[`source entrypoint option flow: source-entrypoint-option-flow.cjs.dev.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
let something = \\"\\";
exports.something = something;
"
`;

exports[`source entrypoint option flow: source-entrypoint-option-flow.cjs.js 1`] = `
"'use strict';

if (process.env.NODE_ENV === \\"production\\") {
  module.exports = require(\\"./source-entrypoint-option-flow.cjs.prod.js\\");
} else {
  module.exports = require(\\"./source-entrypoint-option-flow.cjs.dev.js\\");
}
"
`;

exports[`source entrypoint option flow: source-entrypoint-option-flow.cjs.js.flow 1`] = `
"// @flow
export * from \\"../modules/index.js\\";
"
`;

exports[`source entrypoint option flow: source-entrypoint-option-flow.cjs.prod.js 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: !0
});

let something = \\"\\";

exports.something = something;
"
`;

exports[`source entrypoint option: source-entrypoint-option.cjs.dev.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
let something = \\"\\";
exports.something = something;
"
`;

exports[`source entrypoint option: source-entrypoint-option.cjs.js 1`] = `
"'use strict';

if (process.env.NODE_ENV === \\"production\\") {
  module.exports = require(\\"./source-entrypoint-option.cjs.prod.js\\");
} else {
  module.exports = require(\\"./source-entrypoint-option.cjs.dev.js\\");
}
"
`;

exports[`source entrypoint option: source-entrypoint-option.cjs.prod.js 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: !0
});

let something = \\"\\";

exports.something = something;
"
`;

exports[`two entrypoints where one requires the other entrypoint: dist/common-dependency-two-entrypoints.cjs.dev.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

let identity = x => x;

exports.identity = identity;
"
`;

exports[`two entrypoints where one requires the other entrypoint: dist/common-dependency-two-entrypoints.cjs.js 1`] = `
"'use strict';

if (process.env.NODE_ENV === \\"production\\") {
  module.exports = require(\\"./common-dependency-two-entrypoints.cjs.prod.js\\");
} else {
  module.exports = require(\\"./common-dependency-two-entrypoints.cjs.dev.js\\");
}
"
`;

exports[`two entrypoints where one requires the other entrypoint: dist/common-dependency-two-entrypoints.cjs.prod.js 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: !0
});

let identity = x => x;

exports.identity = identity;
"
`;

exports[`two entrypoints where one requires the other entrypoint: multiply/dist/common-dependency-two-entrypoints.cjs.dev.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var commonDependencyTwoEntrypoints = require('../../dist/common-dependency-two-entrypoints.cjs.dev.js');

let multiply = (a, b) => commonDependencyTwoEntrypoints.identity(a * b);

exports.multiply = multiply;
"
`;

exports[`two entrypoints where one requires the other entrypoint: multiply/dist/common-dependency-two-entrypoints.cjs.js 1`] = `
"'use strict';

if (process.env.NODE_ENV === \\"production\\") {
  module.exports = require(\\"./common-dependency-two-entrypoints.cjs.prod.js\\");
} else {
  module.exports = require(\\"./common-dependency-two-entrypoints.cjs.dev.js\\");
}
"
`;

exports[`two entrypoints where one requires the other entrypoint: multiply/dist/common-dependency-two-entrypoints.cjs.prod.js 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: !0
});

var commonDependencyTwoEntrypoints = require(\\"../../dist/common-dependency-two-entrypoints.cjs.prod.js\\");

let multiply = (a, b) => commonDependencyTwoEntrypoints.identity(a * b);

exports.multiply = multiply;
"
`;

exports[`two entrypoints where one requires the other entrypoint: src/identity.js 1`] = `
"export let identity = x => x;
"
`;

exports[`two entrypoints where one requires the other entrypoint: src/multiply.js 1`] = `
"import { identity } from \\"./identity\\";

export let multiply = (a, b) => identity(a * b);
"
`;

exports[`two entrypoints with a common dependency: dist/chunk-056d83ff.esm.js 1`] = `
"let identity = x => x;

export { identity as a };
"
`;

exports[`two entrypoints with a common dependency: dist/chunk-770164e0.cjs.dev.js 1`] = `
"'use strict';

let identity = x => x;

exports.identity = identity;
"
`;

exports[`two entrypoints with a common dependency: dist/chunk-770164e0.cjs.prod.js 1`] = `
"\\"use strict\\";

let identity = x => x;

exports.identity = identity;
"
`;

exports[`two entrypoints with a common dependency: dist/common-dependency-two-entrypoints.cjs.dev.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var __chunk_1 = require('./chunk-770164e0.cjs.dev.js');

let sum = (a, b) => __chunk_1.identity(a + b);

exports.sum = sum;
"
`;

exports[`two entrypoints with a common dependency: dist/common-dependency-two-entrypoints.cjs.js 1`] = `
"'use strict';

if (process.env.NODE_ENV === \\"production\\") {
  module.exports = require(\\"./common-dependency-two-entrypoints.cjs.prod.js\\");
} else {
  module.exports = require(\\"./common-dependency-two-entrypoints.cjs.dev.js\\");
}
"
`;

exports[`two entrypoints with a common dependency: dist/common-dependency-two-entrypoints.cjs.prod.js 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: !0
});

var __chunk_1 = require(\\"./chunk-770164e0.cjs.prod.js\\");

let sum = (a, b) => __chunk_1.identity(a + b);

exports.sum = sum;
"
`;

exports[`two entrypoints with a common dependency: dist/common-dependency-two-entrypoints.esm.js 1`] = `
"import { a as identity } from './chunk-056d83ff.esm.js';

let sum = (a, b) => identity(a + b);

export { sum };
"
`;

exports[`two entrypoints with a common dependency: multiply/dist/common-dependency-two-entrypoints.cjs.dev.js 1`] = `
"'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var __chunk_1 = require('../../dist/chunk-770164e0.cjs.dev.js');

let multiply = (a, b) => __chunk_1.identity(a * b);

exports.multiply = multiply;
"
`;

exports[`two entrypoints with a common dependency: multiply/dist/common-dependency-two-entrypoints.cjs.js 1`] = `
"'use strict';

if (process.env.NODE_ENV === \\"production\\") {
  module.exports = require(\\"./common-dependency-two-entrypoints.cjs.prod.js\\");
} else {
  module.exports = require(\\"./common-dependency-two-entrypoints.cjs.dev.js\\");
}
"
`;

exports[`two entrypoints with a common dependency: multiply/dist/common-dependency-two-entrypoints.cjs.prod.js 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: !0
});

var __chunk_1 = require(\\"../../dist/chunk-770164e0.cjs.prod.js\\");

let multiply = (a, b) => __chunk_1.identity(a * b);

exports.multiply = multiply;
"
`;

exports[`two entrypoints with a common dependency: multiply/dist/common-dependency-two-entrypoints.esm.js 1`] = `
"import { a as identity } from '../../dist/chunk-056d83ff.esm.js';

let multiply = (a, b) => identity(a * b);

export { multiply };
"
`;

exports[`two entrypoints with a common dependency: src/identity.js 1`] = `
"export let identity = x => x;
"
`;

exports[`two entrypoints with a common dependency: src/multiply.js 1`] = `
"import { identity } from \\"./identity\\";

export let multiply = (a, b) => identity(a * b);
"
`;

exports[`two entrypoints with a common dependency: src/sum.js 1`] = `
"import { identity } from \\"./identity\\";

export let sum = (a, b) => identity(a + b);
"
`;
